samza-test/src/test/java/org/apache/samza/test/startpoint/StartpointTestHarness.java

```diff
@@ -0,0 +1,383 @@
+  private ApplicationConfig applicationConfig3;
+  private ApplicationConfig applicationConfig4;
+  private String testStreamAppName;
+  private String testStreamAppId;
+
+  @Rule
+  public Timeout testTimeOutInMillis = new Timeout(150000, TimeUnit.MILLISECONDS);
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  @Override
+  public void setUp() {
+    super.setUp();
+    String uniqueTestId = UUID.randomUUID().toString();
+    testStreamAppName = String.format("test-app-name-%s", uniqueTestId);
+    testStreamAppId = String.format("test-app-id-%s", uniqueTestId);
+    inputKafkaTopic1 = String.format("test-input-topic1-%s", uniqueTestId);
+    inputKafkaTopic2 = String.format("test-input-topic2-%s", uniqueTestId);
+    inputKafkaTopic3 = String.format("test-input-topic3-%s", uniqueTestId);
+    inputKafkaTopic4 = String.format("test-input-topic4-%s", uniqueTestId);
+    outputKafkaTopic = String.format("test-output-topic-%s", uniqueTestId);
+
+    // Set up stream application config map with the given testStreamAppName, testStreamAppId and test kafka system
+    // TODO: processorId should typically come up from a processorID generator as processor.id will be deprecated in 0.14.0+
+    Map<String, String> configMap =
+        buildStreamApplicationConfigMap(testStreamAppName, testStreamAppId);
+    configMap.put(JobConfig.PROCESSOR_ID, PROCESSOR_IDS[0]);
+    applicationConfig1 = new ApplicationConfig(new MapConfig(configMap));
+    configMap.put(JobConfig.PROCESSOR_ID, PROCESSOR_IDS[1]);
+    applicationConfig2 = new ApplicationConfig(new MapConfig(configMap));
+    configMap.put(JobConfig.PROCESSOR_ID, PROCESSOR_IDS[2]);
+    applicationConfig3 = new ApplicationConfig(new MapConfig(configMap));
+    configMap.put(JobConfig.PROCESSOR_ID, PROCESSOR_IDS[3]);
+    applicationConfig4 = new ApplicationConfig(new MapConfig(configMap));
+
+    ImmutableMap<String, Integer> topicToPartitionCount = ImmutableMap.<String, Integer>builder()
+        .put(inputKafkaTopic1, ZK_TEST_PARTITION_COUNT)
+        .put(inputKafkaTopic2, ZK_TEST_PARTITION_COUNT)
+        .put(inputKafkaTopic3, ZK_TEST_PARTITION_COUNT)
+        .put(inputKafkaTopic4, ZK_TEST_PARTITION_COUNT)
+        .put(outputKafkaTopic, ZK_TEST_PARTITION_COUNT)
+        .build();
+
+    List<NewTopic> newTopics =
+        topicToPartitionCount.keySet()
+            .stream()
+            .map(topic -> new NewTopic(topic, topicToPartitionCount.get(topic), (short) 1))
+            .collect(Collectors.toList());
+
+    assertTrue("Encountered errors during test setup. Failed to create topics.", createTopics(newTopics));
+  }
+
+  @Override
+  public void tearDown() {
+    deleteTopics(
+        ImmutableList.of(inputKafkaTopic1, inputKafkaTopic2, inputKafkaTopic3, inputKafkaTopic4, outputKafkaTopic));
+    SharedContextFactories.clearAll();
+    super.tearDown();
+  }
+
+  @Test
+  public void testStartpoints() throws InterruptedException {
```

Comment:
Can you break this method down into multiple functions: like one for startpoint specific, one for startpoint timestamp, etc. The code in its current state is too lengthy and not legible.

Reply:
Sure good suggestion.